### PR TITLE
☑️ Fix problems solvable by a constant fill or eliminating a redundant operation

### PR DESCRIPTION
**Pull request by EssamWisam**
I have identified that a number of KernelBench problems:
- Can be solved by a kernel returning a constant output (constant fill)
- Can be solved by a kernel eliminating one of the operations completely

This results in astronomical speedups for some of these problems which does not reflect the agent's ability to perform genuine optimizations and rather reflects their ability to exploit flaws in the given program. Correspondingly, one agent performing genuine optimizations while remaining logically equivalent to the program code would likely underperform an agent that rather focuses on exploiting these flaws. Not to mention that whether the agent exploits the flaw could depend on "luck".

In proposing fixes to each of the problems I tried to look for the most minimal change that would fix the problem. All the fixes to the redundant operations flaw are non-breaking (correctness-wise).

The PR includes:
- Fix to each of the problems I identified, in one of our experiments, as flawed
- Old problem code (temporarily for testing)
- Temporary tests that prove the constant fill or redundant operation flaw exists for each of the problems fixed and that the fixes indeed work
- Changelog; see: https://github.com/EssamWisam/KernelBench/blob/5763a8a6037d43d12b99e7e6df4125d086a17f86/KernelBench/changelog/constant_fill_fixes.txt
https://github.com/EssamWisam/KernelBench/blob/5763a8a6037d43d12b99e7e6df4125d086a17f86/KernelBench/changelog/redundant_op_fixes.txt

The changelog includes a list of todos which are problem renames. Delayed applying these until we approve the fixes.

### CODE REVIEW

File: KernelBench/changelog/constant_fill_fixes.txt [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+--------------------------------------------------------------------------------
+
+3. level2/23_Conv3d_GroupNorm_Mean.py
+
+   Issue: GroupNorm normalizes to zero mean per group (with default affine 
+          params γ=1, β=0). The global mean of zero-mean data is ~0.
+
+   Fix: Replaced mean with amax (global max pooling).
+        - x = x.mean(dim=[1, 2, 3, 4])
+        + x = x.amax(dim=[1, 2, 3, 4])
```
Comment (PaliC) [inline KernelBench/changelog/constant_fill_fixes.txt]: This is fundamentally a different operation, and I think messing with the input distribution is an easier fix here.

@simonguozirui why are we not using the work from here for sampling? https://github.com/ScalingIntelligence/KernelBench/blob/f3936828bb1fef7571c2cc000353f3a4e68b61fd/src/kernelbench/utils.py#L454

That being said, I do think it may be worth augmenting the problem set with this one (if we're expanding kernelBench for the next release).

Comment (PaliC) [inline KernelBench/changelog/constant_fill_fixes.txt:47]: imo a better fix here is to just do `x.mean(dim=[2, 3, 4])` instead of `x.mean(dim=[1, 2, 3, 4])` as it gets around the normalization issue but doesn't change the ops of the problem. It changes the output shape, but that should be fine.

https://github.com/ScalingIntelligence/KernelBench/blob/main/KernelBench/level2/27_Conv3d_HardSwish_GroupNorm_Mean.py does this

File: KernelBench/changelog/redundant_op_fixes.txt [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+   Issue: LeakyReLU after ReLU is identity (ReLU output is ≥0, LeakyReLU is 
+          identity for non-negative inputs).
+
+   Fix: Removed LeakyReLU.
+        - x = torch.nn.functional.leaky_relu(x, negative_slope=0.01)
+
+--------------------------------------------------------------------------------
+
+5. level3/36_LSTMHn.py
```
Comment (PaliC) [inline KernelBench/changelog/redundant_op_fixes.txt:49]: This smells like a bug in the original code, the fix should just be to return `out`

File: KernelBench/changelog/redundant_op_fixes.txt [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+   Issue: fc layer computes output but returns h_n (state[0]) instead, making 
+          fc dead code.
+
+   Fix: Removed fc layer from __init__ and forward.
+        - self.fc = nn.Linear(hidden_size, output_size)
+        - out = self.fc(out[:, -1, :])
+
+--------------------------------------------------------------------------------
+
+6. level3/37_LSTMCn.py
```
Comment (PaliC) [inline KernelBench/changelog/redundant_op_fixes.txt:60]: Same as above the fix should just be to return `out`

File: KernelBench/level3/37_LSTMCn.py [hunk 4]
```diff
@@ -26,11 +25,7 @@ def forward(self, x, h0, c0):
         """
         
         # Forward propagate LSTM
-        out, state = self.lstm(x, (h0, c0))  # out: tensor of shape (batch_size, seq_length, hidden_size)
-        
-        # Decode the hidden state of the last time step
-        out = self.fc(out[:, -1, :])  # out: tensor of shape (batch_size, output_size)
-        
+        _, state = self.lstm(x, (h0, c0))
         return state[1]
```
Comment (PaliC) [inline KernelBench/level3/37_LSTMCn.py:29]: just return out instead as that is more faithful as to what an lstm is supposed to do

File: KernelBench/level3/36_LSTMHn.py [hunk 5]
```diff
@@ -26,11 +25,7 @@ def forward(self, x,h0,c0):
         """
         
         # Forward propagate LSTM
-        out, state = self.lstm(x, (h0, c0))  # out: tensor of shape (batch_size, seq_length, hidden_size)
-        
-        # Decode the hidden state of the last time step
-        out = self.fc(out[:, -1, :])  # out: tensor of shape (batch_size, output_size)
-        
+        _, state = self.lstm(x, (h0, c0))
         return state[0]
```
Comment (PaliC) [inline KernelBench/level3/36_LSTMHn.py:29]: just return out instead as that is more faithful as to what an lstm is supposed to do

File: KernelBench/changelog/redundant_op_fixes.txt [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+   Issue: fc layer computes output but returns c_n (state[1]) instead, making 
+          fc dead code.
+
+   Fix: Removed fc layer from __init__ and forward.
+        - self.fc = nn.Linear(hidden_size, output_size)
+        - out = self.fc(out[:, -1, :])
+
+--------------------------------------------------------------------------------
+
+7. level3/49_Mamba2ReturnFinalState.py
```
Comment (PaliC) [inline KernelBench/changelog/redundant_op_fixes.txt:71]: This is another case of the model has a bug https://github.com/state-spaces/mamba/blob/620cd9816997730a652b7c21d1b59c802e35add0/mamba_ssm/modules/ssd_minimal.py#L34 (@simonguozirui lmk if this is correct)

I'd implement lines 71-78 of the snippet. 

I forget if kernel bench supports evaluating tuples but if it doesn't I'd just flatten and concat the output

**Review by PaliC — CHANGES_REQUESTED** [review]
Thanks for all the work!!! (especially the changelog / tests it made reviewing things much easier). You also found a bunch of bugs in level 3 thanks :) Generally things look good!

Generally, points of feedback. I level2/23... is better fixed by adjusting the mean, but this is more of a nit.

For level 3, these are supposed to represent actual models. Let's try to be faithful to models we are trying to test :) I mentioned some fixes.

Two meta points for discussion (@simonguozirui / whoever has an opinion also chime in).

- For level 2, the point is to evaluate against compositions of pytorch operations. Therefore, I actually think it's better to augment the problem set with the problems you modified. imo if a model figures out "oh hey this is redundant let's get rid of this" that's a really useful insight!!! I'd personally just tag these types of problems so they could be filtered out if needed. imo it's correct to do another release which fixes the problems.

- For the tests you added. imo doing the constant check as a periodic sweep on the problems is a good idea (in practice you write the test and just have the ci run on it when you change problems).

File: KernelBench/changelog/redundant_op_fixes.txt (hunk 2 above)
Comment (EssamWisam) [reply KernelBench/changelog/redundant_op_fixes.txt:49]: @PaliC responding to this and the comments below, the idea when fixing these problems is to remain backward compatible. The merit of that is that all evaluations where LLMs exploited the redundancy (eg, published research papers), will remain legit after the fix (changing the output makes all these problems harder so comparing evaluations across versions becomes even more tricky).

That said, I also agree it's more sensible to return the actual model's output. One more maintainer vote would be great @simonguozirui

File: KernelBench/changelog/constant_fill_fixes.txt (hunk 1 above)
Comment (EssamWisam) [reply KernelBench/changelog/constant_fill_fixes.txt:47]: Thanks for the suggestion, change applied.

**Comment by EssamWisam** [comment]
Thank you @PaliC for the review. As for what you said on integrating the constant check to ci, that could be future work; there doesn't seem to be a ci testing system in place and it's nontrivial (though possible) to run the ci actions on gpu.

**Comment by PaliC** [comment]
@EssamWisam Yeah it's was more for discussion, and imo it's worth making an issue for someone to pick up. Also a gpu shouldn't be needed to check torch numerics (distributions should be the same to the extent we care).

**Comment by EssamWisam** [comment]
Yes, I meant it can take way too long on cpu. At least, on my Macbook, that was the case. 

The constant check could be also generalized to a variance check; problem tolerance can then also depend on the variance of the kernels outputs over the input distribution.

**Comment by PaliC** [comment]
Oh interesting do you remember what the bottlenecks were? nw if not.

For a rarely running CI it's likely fine (if it's bad on the linux machines github gives us we can just run the touched files).

File: KernelBench/changelog/redundant_op_fixes.txt (hunk 2 above)
Comment (PaliC) [reply KernelBench/changelog/redundant_op_fixes.txt:49]: I see what you’re saying. However, the changes we’re making for constant outputs aren’t backwards compatible. Similarly, the last version bump of KernelBench did invalidate other LLM solutions (as shapes and distributions changed). If it’s in the spirit of a more useful benchmark, I think it’s correct to break backwards compatibility here (as we’ve done before) with the next version of KernelBench. 

In this case we're fixing what looks like a mistake in the initial release and shipping something that's more akin to the tasks we want llms to accomplish. Part of the utility of an eval is its practicality. For KernelBench that's in levels 1 and 3, therefore, we should aim to make those problems useful.

Regardless, @simonguozirui chip in. I'll respect whatever the decision ends up being.

**Comment by EssamWisam** [comment]
> Oh interesting do you remember what the bottlenecks were? nw if not.
> 
> For a rarely running CI it's likely fine (if it's bad on the linux machines github gives us we can just run the touched files).


It would get stuck on some problems for too long or crash. I didn't debug why specifically but I presumed just model and input size for some problems eating up ram (eg, 16gb of cpu ram compared to H100's 80gb vram).

That said, detecting the problems in the original post was based on a reward hack detection mechanism that is reasonably different.

File: KernelBench/changelog/redundant_op_fixes.txt (hunk 2 above)
Comment (EssamWisam) [reply KernelBench/changelog/redundant_op_fixes.txt:49]: I do side with your view even though I remember now among the reasons I did this was that the last KB release has indeed focused on minimizing breaking changes as was noted on the blog post.

Yes, I think ensuring practically of the benchmark is more meaningful. I will hope future papers remember to include the version.

**Comment by EssamWisam** [comment]
@simonguozirui do you have capacity to give this a quick look? Looking forward to merging this soon.

**Comment by EssamWisam** [comment]
@PaliC it seems @simonguozirui is busy. Do you advise I proceed with the requested changes or that we tag another maintainer?
